src/ILOPoolSale.sol

```diff
@@ -0,0 +1,185 @@
+import { PoolAddress } from './libraries/PoolAddress.sol';
+import { FullMath } from '@uniswap/v3-core/contracts/libraries/FullMath.sol';
+
+contract ILOPoolSale is ILOPoolBase, IILOPoolSale, ILOWhitelist {
+    bool public override CANCELLED;
+    uint64 public override SALE_START;
+    uint64 public override SALE_END;
+    uint256 public override MIN_RAISE;
+    uint256 public override MAX_RAISE;
+    uint256 public override TOTAL_RAISED;
+
+    VestingConfig private _vestingConfig;
+
+    modifier onlyProjectAdmin() override {
+        IILOManager.Project memory _project = IILOManager(MANAGER).project(
+            PROJECT_ID
+        );
+        require(msg.sender == _project.admin, 'UA');
+        _;
+    }
+
+    modifier duringSale() {
+        require(SALE_START <= block.timestamp, 'BS');
+        require(SALE_END > block.timestamp, 'ES');
+        _;
+    }
+
+    modifier whenCancelled() {
+        require(CANCELLED, 'NOT_CANCELLED');
+        _;
+    }
+
+    modifier whenNotCancelled() {
+        require(!CANCELLED, 'CANCELLED');
+        _;
+    }
+
+    function initialize(InitParams calldata params) external override {
+        _initialize(params.poolParams);
+        require(params.poolParams.vestingConfigs.length == 1, 'VC');
+        _vestingConfig = params.poolParams.vestingConfigs[0];
+        (SALE_START, SALE_END, MIN_RAISE, MAX_RAISE) = (
+            params.saleParams.start,
+            params.saleParams.end,
+            params.saleParams.minRaise,
+            params.saleParams.maxRaise
+        );
+    }
+
+    function buy(
+        uint256 raiseAmount,
+        address recipient
+    ) external override duringSale whenNotCancelled returns (uint256 tokenId) {
+        require(raiseAmount > 0, 'ZA');
+        // check if raise amount over capacity
+        uint256 totalRaised = TOTAL_RAISED;
+        require(MAX_RAISE - totalRaised >= raiseAmount, 'MR');
+        totalRaised += raiseAmount;
+        TOTAL_RAISED = totalRaised;
+
+        // if investor already have a position, just increase raise amount and liquidity
+        // otherwise, mint new nft for investor and assign vesting schedules
+        if (balanceOf(recipient) == 0) {
+            _mint(recipient, (tokenId = _nextId++));
+            _positionVests[tokenId].schedule = _vestingConfig.schedule;
+        } else {
+            tokenId = tokenOfOwnerByIndex(recipient, 0);
+        }
+
+        Position storage _position = _positions[tokenId];
+
+        uint256 _allocation = allocation(recipient);
+        // we need to check `_position.raiseAmount < _allocation`
+        // because project admin can change whitelist during sale.
+        // so, technicaly, `_position.raiseAmount` can bigger than `_allocation`
+        // second reason is to avoid overflow of subtraction.
+        require(
+            _position.raiseAmount < _allocation &&
+                raiseAmount <= _allocation - _position.raiseAmount,
```

Comment:
raiseAmount <= _allocation - _position.raiseAmount is enough

and `raiseAmount + _position.raiseAmount <= _allocation` is more neutral

Reply:
It had comment to explain why we need to check both and subtraction check not addition